Brightness adaptive scanline support

wickerwaka opened the pull request:
Interpolate between two sets of filter coefficients based on the luminosity of the source pixel. Intended to be used to simulate the widening of CRT phosphor glow as the intensity of the beam increases.

The pixel luminosity is approximated by .25R + .5G + .25B. The resulting luminance value is used to interpolate between the two selected coefficient sets. The interpolation requires 4 DSP blocks operating in Sum of Two 18x18 mode. In order to reduce the amount of DSP resources being used only a single dimension is supported, you can have adaptive scanlines in the vertical or horizontal direction, but not both.

The `PolyFetch` process is responsible for fetching all polyphase coefficients and performing the interpolation. It takes 4 cycles to interpolate once the luminance value is ready, this requires lengthening the pipeline of both the horizontal and vertical stages. Since the color data is required both to calculate the luminance and to apply the filter and there is at least 5 cycles between those two things, the 96-bits of pixel data needs to be copied forward through each stage. In the vertical stage this was being inferred as a shift register in block ram and due to the size it was using 3 block rams to do it. To reduce the block ram usage the line required to calculate the luminance data is read using a different address than the other three lines so only 24-bits of color data needs to be copied through all of the stages.

The separate nearest neighbor filtering support has been removed and it is now done as a special case of polyphase filtering. It's worth considering doing the same with bilinear and bicubic filtering, even though they are not currently used by anything, keeping their code in sync with other changes is complex.

Resource usage changes across three test platforms is as follows:
|     | Genesis | SNES | PSX |
| --- | --- | --- | --- |
| ALM | +92 | +442 | +172 |
| DSP | +3 | +3 | +3 |
| Block RAM | +2 | +3 | +2 |

4 DSP blocks are used for the interpolation, however the net increase from this change is only 3 because I replaced an unnecessary DSP multiply when calculating the `o_hburst` size.

There is only one explicit block ram being used in this change, that is the ram for the second set of polyphase coefficients. Two additional block rams are used for inferred shift registers. One for `o_vpix_inner`, which is the 24-bits of color data that needs to be preserved through the vertical stages and 1 for the `o_dcptv` array, which increased in length due to extra horizontal stages and seems to have passed a threshold.

Net ALM and register usage varies a lot across the different cores.

Added `MISTER_DISABLE_ADAPTIVE` and `MISTER_DOWNSCALE_NN` macros to disable the adaptive filters and bilinear downscaling for cores that need extra ALM or DSP resources.
 
| Resource | Base | New-Base | Adaptive | NN | Adaptive/NN |
| --- | --- | --- | --- | --- | --- |
| ALM | 36313 | 36411 (+98) | 36485 (+172) | 36267 (-46) | 36355 (+42) |
| DSP | 111 | 110 (-1) | 112 (+1) | 104 (-7) | 108 (-3) |
| Block Ram | 267 | 268 (+1) | 269 (+2) | 268 (+1) | 269 (+2) |

* "Base" is the current PSX build
* "New-Base" is this build with but adaptive filters disabled (`MISTER_DISABLE_ADAPTIVE=1`)
* "Adaptive" is with adaptive filters them enabled
* "NN" is adaptive-disabled and downsampling forced to nearest neighbor (`MISTER_DOWNSCALE_NN=1`)
* "Adaptive/NN" is adaptive enabled and downsampling forced to nearest neighbor. (`MISTER_DISABLE_ADAPTIVE=1` and `MISTER_DOWNSCALE_NN=1`)

@Grabulosaure @ghogan42

wickerwaka commented:
Requires https://github.com/MiSTer-devel/Main_MiSTer/pull/520

sorgelig commented:
I know this opinion is not popular but i think requirement of integer scaling is main distracting point of the whole feature. Having borders all around the screen is ugly and not acceptable for me personally. But my discussion is not about like it or not. What i don't understand that switching to 64 phases basically gained nothing in scaling quality. I even tried 256 phase and didn't find noticeable difference from 16 phase (well, it's very very tiny difference at the tolerance level) which shouldn't be the case. I believe 64 phases (and especially 256) should provide much better non-integer scale. I don't know how my iPad display with Chinese LCD driver board scales, but it scales ANY resolution to 1536p so good so you won't differentiate it from integer scaling. I even tried video from MiSTer 1080p with scanlines and those scanlines on 1536p were perfect like integer scaling was applied. Remember it was scaled from 1080p to 1536p by display. I believe 64 phases should scale significantly better than 16 but there must be some other limiting factor in ascal preventing it. Since you already worked with ascal code, may be you can check this aspect. I hope one day user won't need to choose between scanlines and full screen.

wickerwaka commented:
I wonder if there is a precision/accuracy issue with the fractional values (`o_vfrac` and `o_hfrac`)?

ghogan42 commented:
I can't say if there is a problem with the scaler in terms of precision.  But if you are going to compare filter scanlines vs scandoubler scanlines scaled by mister vs scandoubler scanlines scaled by a TV you have to be careful to compare the right things.

Scandoubler scanlines are applied with an offset compared to filter scanlines. And the "brightness boosted" scanlines will always be worse quality than "normal" scanlines.

I've made a quick comparison of scandoubler 50% scanlines with the scanlines_50 filter scanlines. If anyone wants me to do it I can do the same for 75% and 25% scanlines as well. 75% may show more differences.

Be sure to view at 100% size. Or 200%. 300% etc.

![scandoubler_and_filter_comparison](https://user-images.githubusercontent.com/35882078/153681174-e7c0974c-33e7-45b0-a0b3-2e0783e587bd.png)

sorgelig commented:
Well, after introduction of separate filter for scandoubler scanlines it became possible to have a decent scanlines without integer scaling using more smooth filter. But my concern is about difference between 16 and 64 phases. The pic above will hardly distinguished using 16 phases instead.

wickerwaka commented:
![64_vs_16](https://user-images.githubusercontent.com/6377250/153738241-0fe8da6c-3291-4830-b846-4c08f2789d98.png)

64 vs 16 phase comparison with non-integer scaling to 1080p. To my eye the 16 phase version looks pretty bad, but I'd happily play the 64 phase version. Scaling up to 1536 lines may benefit from even more phases, but I don't have an ipad display or anything that can capture at that resolution.

sorgelig commented:
Well, from this screenshot difference is visible. 
How about adaptive scanlines in non-integer scaling?
I don't see adaptive filters yet, so can't test it myself.

Grabulosaure commented:
There is no rounding in hfrac/vfrac, so sometimes an off-by-one coefficient is selected when using non-integer scaling.
Raising number of phases makes that error less significant.

wickerwaka commented:
![64vs256_horizontal](https://user-images.githubusercontent.com/6377250/154812376-afd1a2fe-54ec-440e-ae15-955204fc7506.png)
![64vs256_vertical](https://user-images.githubusercontent.com/6377250/154812379-44657565-6f98-4235-ab7d-a3d6ffaac7b8.png)

@ghogan42 provided by with some 256 phase filters and I modified ascal to support them. Comparison shots are from non-integer scaling on the genesis core at 1080p. I added `vscale_border=11` since that scaling factor made the issues most apparent at 64 phases.

This can be done pretty cheaply on the vertical axis, however doing it on the horizontal access requires adding an additional stage to do two additional bits of precision in the division.

sorgelig commented:
if doing 256 phases then both X and Y should support it.
